unicon_backend/evaluator/tasks/programming/steps.py

```diff
@@ -91,10 +90,58 @@ def label(self) -> str:
         return self.id.split(".", 2)[-1]
 
 
+class Operator(StrEnum):
+    LESS_THAN = "<"
+    EQUAL = "="
+    GREATER_THAN = ">"
+
+
+class Comparison(CustomSQLModel):
+    operator: Operator
+    value: Any
+
+    @model_validator(mode="after")
+    def check_value_type(self) -> Self:
+        """For < and >, check the operator can be compared (i.e. primitive)"""
+        if self.operator == Operator.EQUAL:
+            return self
+        if not isinstance(self.value, PrimitiveData):
+            raise ValueError(f"Invalid comparison value {self.value} for operator {self.operator}")
+        return self
+
+    def compare(self, actual_value: Any):
+        try:
+            match self.operator:
+                case Operator.EQUAL:
+                    return actual_value == self.value
+                case Operator.LESS_THAN:
+                    return actual_value < self.value
+                case Operator.GREATER_THAN:
+                    return actual_value > self.value
+                case _:
+                    return False
+        except:
+            # if there was an exception, the type returned was incorrect.
+            # so return False.
+            return False  # noqa: B012
+
+
+class OutputSocket(StepSocket):
+    user_label: str | None = None
```

Comment:
it was originally named `label`, but now that we are putting these fields at the socket level, do we want to use the label in the socket id definition (e.g. `DATA.IN.<label>`) or should this be a separate field

current stopgap is to name it `user_label` to avoid name collision

Reply:
currently, i think it is better to separate it so that the admin can display anything to the user, including duplicate labels/whatever future constraints we have on socket names.

unrelated problem i thought of while thinking about this: do spaces break our program because we may use that in the variable name? we only replace "." to "_"

🤔

Reply:
yeap it break things if there is a space in `Socket::id`, so if we were to expose edit the label component of the socket id to users, we will need to catch this case